--- OpenBLAS-0.2.20/kernel/x86_64/daxpy.c~	2017-07-24 04:03:35.000000000 +0000
+++ OpenBLAS-0.2.20/kernel/x86_64/daxpy.c	2018-01-01 16:32:23.567020976 +0000
@@ -38,7 +38,7 @@
 #elif defined(PILEDRIVER)
 #include "daxpy_microk_piledriver-2.c"
 #elif defined(HASWELL) || defined(ZEN)
-#include "daxpy_microk_haswell-2.c"
+#include "daxpy_microk_sandy-2.c"
 #elif defined(SANDYBRIDGE)
 #include "daxpy_microk_sandy-2.c"
 #endif
