--- OpenBLAS-0.2.16/driver/level3/level3_thread.c~	2016-03-15 18:49:10.000000000 +0000
+++ OpenBLAS-0.2.16/driver/level3/level3_thread.c	2016-07-24 02:23:27.754365318 +0000
@@ -608,6 +608,15 @@
   while (m > 0){
 
     width  = blas_quickdivide(m + nthreads - num_cpu_m - 1, nthreads - num_cpu_m);
+    
+    /* prefer widths of multiples of 4 or 8 except for the last block*/
+    if (nthreads - num_cpu_m > 1 && width > 8) {
+       if (width < 64)
+         width = width & ~3;
+       else
+         width = width & ~7;
+    }
+    if (m - width < 8) width = m;
 
     m -= width;
     if (m < 0) width = width + m;
--- OpenBLAS-0.2.16/driver/level3/level3_thread.c~	2016-07-24 02:23:27.000000000 +0000
+++ OpenBLAS-0.2.16/driver/level3/level3_thread.c	2016-07-24 02:26:32.145187880 +0000
@@ -659,6 +659,14 @@
 
       width  = blas_quickdivide(n + nthreads - num_cpu_n - 1, nthreads - num_cpu_n);
 
+    /* prefer widths of multiples of 4 or 8 except for the last block*/
+    if (nthreads - num_cpu_n > 1 && width > 8) {
+       if (width < 64)
+         width = width & ~3;
+       else
+         width = width & ~7;
+    }
+    if (n - width < 8) width = n;
       n -= width;
       if (n < 0) width = width + n;
 
