--- OpenBLAS-0.2.16/kernel/x86_64/KERNEL.HASWELL~	2016-07-17 03:50:30.000000000 +0000
+++ OpenBLAS-0.2.16/kernel/x86_64/KERNEL.HASWELL	2016-07-17 03:50:40.094572128 +0000
@@ -31,9 +31,9 @@
 
 STRMMKERNEL    =  sgemm_kernel_16x4_haswell.S
 SGEMMKERNEL    =  sgemm_kernel_16x4_haswell.S
-SGEMMINCOPY    =  ../generic/gemm_ncopy_16.c
+SGEMMINCOPY    =  ../generic/gemm_ncopy_1.c
 SGEMMITCOPY    =  ../generic/gemm_tcopy_16.c
-SGEMMONCOPY    =  ../generic/gemm_ncopy_4.c
+SGEMMONCOPY    =  ../generic/gemm_ncopy_1.c
 SGEMMOTCOPY    =  ../generic/gemm_tcopy_4.c
 SGEMMINCOPYOBJ =  sgemm_incopy$(TSUFFIX).$(SUFFIX)
 SGEMMITCOPYOBJ =  sgemm_itcopy$(TSUFFIX).$(SUFFIX)
--- OpenBLAS-0.2.16/kernel/generic/gemm_ncopy_1.c~	2016-03-15 18:49:10.000000000 +0000
+++ OpenBLAS-0.2.16/kernel/generic/gemm_ncopy_1.c	2016-07-17 03:51:34.949900181 +0000
@@ -54,26 +54,7 @@
       a_offset1 = a_offset;
       a_offset += lda;
 
-      i = (m >> 3);
-
-      if (i > 0){
-	do {
-	  *(b_offset + 0) = *(a_offset1 + 0);
-	  *(b_offset + 1) = *(a_offset1 + 1);
-	  *(b_offset + 2) = *(a_offset1 + 2);
-	  *(b_offset + 3) = *(a_offset1 + 3);
-	  *(b_offset + 4) = *(a_offset1 + 4);
-	  *(b_offset + 5) = *(a_offset1 + 5);
-	  *(b_offset + 6) = *(a_offset1 + 6);
-	  *(b_offset + 7) = *(a_offset1 + 7);
-	  a_offset1 += 8;
-	  b_offset  += 8;
-	  i --;
-	} while (i > 0);
-      }
-
-      i = (m & 7);
+      i = m;
-
       if (i > 0){
 	do {
 	  *(b_offset + 0) = *(a_offset1 + 0);
--- OpenBLAS-0.2.16/kernel/x86_64/KERNEL.SANDYBRIDGE~	2016-03-15 18:49:10.000000000 +0000
+++ OpenBLAS-0.2.16/kernel/x86_64/KERNEL.SANDYBRIDGE	2016-07-17 04:12:01.437039211 +0000
@@ -25,9 +25,9 @@
 ZAXPYKERNEL = zaxpy.c
 
 SGEMMKERNEL    =  sgemm_kernel_16x4_sandy.S
-SGEMMINCOPY    =  ../generic/gemm_ncopy_16.c
+SGEMMINCOPY    =  ../generic/gemm_ncopy_1.c
 SGEMMITCOPY    =  ../generic/gemm_tcopy_16.c
-SGEMMONCOPY    =  ../generic/gemm_ncopy_4.c
+SGEMMONCOPY    =  ../generic/gemm_ncopy_1.c
 SGEMMOTCOPY    =  ../generic/gemm_tcopy_4.c
 SGEMMINCOPYOBJ =  sgemm_incopy$(TSUFFIX).$(SUFFIX)
 SGEMMITCOPYOBJ =  sgemm_itcopy$(TSUFFIX).$(SUFFIX)
